src/modules/sync/diff/combinators.ts

```diff
@@ -0,0 +1,189 @@
+import { zip } from "../../../utils/array.js";
+import { apply } from "../../../utils/function.js";
+import { PatchOperation } from "../types/diffModel.js";
+
+export type Handler<Source, Target> = (sourceValue: Source, targetValue: Target) => ReadonlyArray<PatchOperation>;
+export type ContextfulHandler<Context, Source, Target> = Readonly<
+  { contextfulHandler: (context: Context) => Handler<Source, Target> }
+>;
+
+/**
+ * Create patch operations for changed properties inside the object.
+ * Provide handlers for all the individual properties of the object.
+ */
+export const makeObjectHandler = <Source extends object, Target extends object>(
+  innerHandlers: Omit<
+    {
+      readonly [k in keyof Source & keyof Target]-?:
+        | Handler<Source[k], Target[k]>
+        | ContextfulHandler<Readonly<{ source: Source; target: Target }>, Source[k], Target[k]>;
+    },
+    "id" | "codename" | "external_id"
+  >,
+): Handler<Source, Target> =>
+(sourceValue, targetValue) => {
+  type IntersectKey = keyof Source & keyof Target;
+
+  return (Object.entries(innerHandlers) as [
+    IntersectKey & string,
+    | Handler<Source[IntersectKey], Target[IntersectKey]>
+    | ContextfulHandler<Readonly<{ source: Source; target: Target }>, Source[IntersectKey], Target[IntersectKey]>,
+  ][])
+    .flatMap(([key, someHandler]) => {
+      const handler = typeof someHandler === "function"
+        ? someHandler
+        : someHandler.contextfulHandler({ source: sourceValue, target: targetValue });
+
+      return handler(sourceValue[key], targetValue[key])
+        .map(prefixOperationPath(key));
+    });
+};
+
+/**
+ * Replaces the whole object if any of its properties are different. (===) is used for comparison.
+ *
+ * @param customComparers - use custom comparers for individual properties. The comparers should return true when the properties are the same and false otherwise.
+ *
+ * @param transformBeforeReplace - provide to transform the source object for the patch operation that will replace the target
+ */
+export const makeLeafObjectHandler = <Source extends object, Target extends object>(
+  customComparers: Omit<
```

Comment:
It can, but I thought that this way people will be less likely to just forget that the parameter is there. An the fact that `{}` is not very long overhead. What do you think?